### PR TITLE
chore(schemas, datadog_logs sink): add schema requirements

### LINKED ISSUE #11306: Add semantic meaning functions that change definition, passing through remap transform

### LINKED ISSUE #11308: Add schema requirement to Datadog Log sink

### LINKED ISSUE #11309: Hook up schema requirements to topology, preventing vector from booting if needed, hide behind new disabled by default global option

### PR DESCRIPTION
**Pull request by JeanMertz**
Follow-up to #11505.

This PR integrates schema validation into the topology. It touches a bunch of earlier landed code that needed updating for the entire schema flow to work as expected.

closes #11306
closes #11308
closes #11309

### CODE REVIEW

File: src/sinks/datadog/logs/sink.rs
Comment (JeanMertz) [inline src/sinks/datadog/logs/sink.rs:267]: We should remove this duplication by integrating this into the `LogRequestBuilder` as a generic type.

File: src/sinks/datadog/logs/sink.rs
Comment (pablosichert) [inline src/sinks/datadog/logs/sink.rs:106]: Note that this will remove the `encoding.{only_fields,except_fields,timestamp_format}` options and therefore be a breaking change.

Comment (pablosichert) [inline src/sinks/datadog/logs/sink.rs:106]: Note that this will remove the `encoding.{only_fields,except_fields,timestamp_format}` options and therefore be a breaking change.

File: src/sources/datadog/agent/tests.rs
Comment (pablosichert) [inline src/sources/datadog/agent/tests.rs:354]: Do we need the `dbg!` invocation here?

Comment (pablosichert) [inline src/sources/datadog/agent/tests.rs:354]: Do we need the `dbg!` invocation here?

File: src/topology/schema.rs
Comment (pablosichert) [inline src/topology/schema.rs:569]: I think I'd remove second pipeline as it doesn't affect the thing to be asserted

Comment (pablosichert) [inline src/topology/schema.rs:569]: I think I'd remove second pipeline as it doesn't affect the thing to be asserted

**Review by pablosichert — COMMENTED** [review]
Really exciting work, @JeanMertz!

A couple of small remarks/questions, and one blocking comment that would introduce a breaking change.

Looks very good overall though and great test coverage.

File: lib/vector-core/src/schema/definition.rs
Comment (JeanMertz) [inline lib/vector-core/src/schema/definition.rs:49]: TODO: this should return `Valid` if both pointers point to the same path.

File: lib/vector-core/src/schema/definition.rs
Comment (JeanMertz) [reply lib/vector-core/src/schema/definition.rs:244]: It behaves differently, but I intend to fix that. See also: https://github.com/vectordotdev/vector/pull/11743/files#r873642818

Specifically, merging two "meaning pointers" currently _always_ returns an invalid pointer, but it shouldn't if both meanings point to the same field (which is what this case covers, which is why tests are passing, even though the `merge` function is technically incorrect).

File: lib/vector-core/src/schema/definition.rs
Comment (JeanMertz) [inline lib/vector-core/src/schema/definition.rs:663]: TODO: add test `same meaning, pointing to same path`

File: lib/vector-core/src/schema/requirement.rs
Comment (JeanMertz) [inline lib/vector-core/src/schema/requirement.rs:105]:
```suggestion
                    // Get the kind at the path for the given semantic meaning. If no kind is found, we set the
```

File: lib/vector-core/src/schema/requirement.rs
Comment (JeanMertz) [inline lib/vector-core/src/schema/requirement.rs:115]:
```suggestion
                        // We found a field matching the defined semantic meaning, but its kind does not
```

File: lib/vector-core/src/schema/requirement.rs
Comment (JeanMertz) [reply lib/vector-core/src/schema/requirement.rs:337]: I'm uncertain if I quite follow what your thoughts are here.

Essentially, this is the same as other type checks, but in this case, the sink defines the field as optional, but since the definition has the field set as present, but as the wrong type, it's an error.

What gave you pause in questioning whether this should be an error?

**Comment by binarylogic** [comment]
@JeanMertz just curious, what are we waiting for to merge this PR? Reviews?

File: src/topology/schema.rs
Comment (pablosichert) [reply src/topology/schema.rs:554]: Does the `source-bar` serve any specific purpose here or can we remove it?

To me it looks like it was copied from the `multiple inputs, sources with schema` test case without being very important here.

**Review by pablosichert — APPROVED** [review]
Overall, this looks in a good shape to me 👍

Some refining questions:

File: lib/vector-core/src/schema/requirement.rs
Comment (JeanMertz) [reply lib/vector-core/src/schema/requirement.rs:336]: It's the former, see also the description of the `optional_meaning` method.

File: src/sinks/datadog/logs/sink.rs
Comment (JeanMertz) [reply src/sinks/datadog/logs/sink.rs:291]: Follow-up. I'll create an issue before merging 👍

**Pull request merged**
